### ISSUE TITLE
How to manually replicate/reproduce/obtain the sha256 hash specified in Nix with fetchgit or fetchFromGitHub?

### ISSUE THREAD

**Issue by hydramo** [NixOS/nix#1880]
At https://nixos.org/nixpkgs/manual/ we see this example:

    src = fetchgit {
      url = "git://github.com/NixOS/nix.git";
      rev = "1f795f9f44607cc5bec70d1300150bfefcef2aae";
      sha256 = "1cw5fszffl5pkpa6s6wjnkiv6lm5k618s32sp60kvmvpy7a2v9kg";
    }
    
    src = fetchgit {
      url = "https://github.com/NixOS/nix.git";
      rev = "1f795f9f44607cc5bec70d1300150bfefcef2aae";
      sha256 = "1cw5fszffl5pkpa6s6wjnkiv6lm5k618s32sp60kvmvpy7a2v9kg";
    }
    
    src = fetchFromGitHub {
      owner = "NixOS";
      repo = "nix";
      rev = "1f795f9f44607cc5bec70d1300150bfefcef2aae";
      sha256 = "04yri911rj9j19qqqn6m82266fl05pz98inasni0vxr1cf1gdgv9";
    }

How do I obtain these hashes myself? I tried the commands

    $ nix-prefetch-url https://github.com/NixOS/nix/archive/1f795f9f44607cc5bec70d1300150bfefcef2aae.tar.gz
    downloading ‘https://github.com/NixOS/nix/archive/1f795f9f44607cc5bec70d1300150bfefcef2aae.tar.gz’... [13/589 KiB, 3.3 KiB/s]
    path is ‘/nix/store/4hrp5jv1rv51lcwf097hbcjjksnwpwi6-1f795f9f44607cc5bec70d1300150bfefcef2aae.tar.gz’
    1311maadkx2xq2ll9apfsyjq5gy5q874z4db7l2riz7vqpx1bxhs
    
    $ nix-prefetch-url https://github.com/NixOS/nix/archive/1f795f9f44607cc5bec70d1300150bfefcef2aae.zip
    downloading ‘https://github.com/NixOS/nix/archive/1f795f9f44607cc5bec70d1300150bfefcef2aae.zip’... [306/835 KiB, 146.0 KiB/s]
    path is ‘/nix/store/gy1zkg71mrlpf5lh8ypl1kg8d2ps6m05-1f795f9f44607cc5bec70d1300150bfefcef2aae.zip’
    1b3vnxfpidwpmgh9dlpmhvrrkkgn3a0r56bv175y6hzfhnx4s5kn

but as you can see, both hashes are completely different from any of the above two.

So how do I actually replicate/reproduce/obtain the sha256 hash specified in Nix with fetchgit or fetchFromGitHub?

**Comment by vcunat** [comment]
It's `--unpack`, as documented for `nix-prefetch-url`.
```
$ nix-prefetch-url --unpack https://github.com/NixOS/nix/archive/1f795f9f44607cc5bec70d1300150bfefcef2aae.tar.gz

unpacking...
path is '/nix/store/74r3qzg6slx4x9lc4v1yvrr7vzjqig3m-1f795f9f44607cc5bec70d1300150bfefcef2aae.tar.gz'
1i2yxndxb6yc9l6c99pypbd92lfq5aac4klq7y2v93c9qvx2cgpc

$ nix-prefetch-url --unpack https://github.com/NixOS/nix/archive/1f795f9f44607cc5bec70d1300150bfefcef2aae.zip
unpacking...
path is '/nix/store/560ws6c0pxr1g67hqpm7w490y1045l31-1f795f9f44607cc5bec70d1300150bfefcef2aae.zip'
1i2yxndxb6yc9l6c99pypbd92lfq5aac4klq7y2v93c9qvx2cgpc
```

The hash `04yri911rj9j19qqqn6m82266fl05pz98inasni0vxr1cf1gdgv9` is wrong.  The `fetchFromGitHub` call above only works with the hash output above (tested now).

**Comment by hydramo** [comment]
Your command is not matching the stuff in fetchFromGitHub. Here is a real example (as opposed to example hardcoded in docs): https://github.com/NixOS/nixpkgs/blob/a78a1d61b56405aeaf04c173e4e3eebbc7cc52b8/pkgs/applications/altcoins/bitcoin-unlimited.nix

If we look into https://github.com/bitcoinunlimited/bitcoinunlimited, we see that the tag `v1.0.2.0` corresponds to commit `9fcdbf50585744696c74c162398cea4f026efda4`.

If we try your command:

    $ nix-prefetch-url --unpack https://github.com/bitcoinunlimited/bitcoinunlimited/archive/9fcdbf50585744696c74c162398cea4f026efda4.tar.gz
    ...
    0v8gkjwl9ivqkq2gj03gsw4syirwnmgb2mrvix7700cnb3xghzbn

That's a totally different hash than that in the `.nix` file above.

Any ideas?

**Comment by vcunat** [comment]
You found another bad expression :-D
```
nix-build -E 'with import <nixpkgs> {}; 
                        fetchFromGitHub {
                            owner = "bitcoinunlimited";
                            repo = "bitcoinunlimited";
                            rev = "v1.0.2.0";
                            sha256 = "17cmy0s1908s2rqs0zwr05f3541nqm2pg08n2xn97g2k3yimdg5q";
                        }'
these derivations will be built:
  /nix/store/hp5hbwz4i9mqazxbcgk2kl7kxvmimqml-bitcoinunlimited-v1.0.2.0-src.drv
building path(s) ‘/nix/store/vliwfr87qw9rgmjxy6rfz0zcynazfv6z-bitcoinunlimited-v1.0.2.0-src’

trying https://github.com/bitcoinunlimited/bitcoinunlimited/archive/v1.0.2.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   143    0   143    0     0    143      0 --:--:-- --:--:-- --:--:--   255
100 6932k  100 6932k    0     0   173k      0  0:00:40  0:00:40 --:--:--  113k
unpacking source archive /tmp/nix-build-bitcoinunlimited-v1.0.2.0-src.drv-0/v1.0.2.0.tar.gz
output path ‘/nix/store/vliwfr87qw9rgmjxy6rfz0zcynazfv6z-bitcoinunlimited-v1.0.2.0-src’ has r:sha256 hash ‘0v8gkjwl9ivqkq2gj03gsw4syirwnmgb2mrvix7700cnb3xghzbn’ when ‘17cmy0s1908s2rqs0zwr05f3541nqm2pg08n2xn97g2k3yimdg5q’ was expected
error: build of ‘/nix/store/hp5hbwz4i9mqazxbcgk2kl7kxvmimqml-bitcoinunlimited-v1.0.2.0-src.drv’ failed
```

**Comment by vcunat** [comment]
The expected hash ^^ is exactly what `nix-prefetch-url --unpack` returned.

**Comment by andrewchambers** [comment]
nix-prefetch-git

**Comment by snawaz** [comment]
`nix-hash` is what you probably need: https://www.mankier.com/1/nix-hash

Use this command to compute the sha256 to be used with `fetchurl` or such: 

```bash
$ nix-hash --flat --base32 --type sha256 your-file.zip
```

**Comment by jhillyerd** [comment]
This should be added to nixpkgs documentation near https://nixos.org/nixpkgs/manual/#chap-pkgs-fetchers

**Comment by kevincox** [comment]
I would be really happy if we could add a fake hash type like `u`, `_` or `unknown` so that I don't need to remember how many 0s I need for a hash that fools nix enough that it tries to download the thing to report the real hash.

**Comment by zarkone** [comment]
good point. For me it would be fine if it would not check the length at all, but I'm afraid it could be harmful to some another parts of system I'm not aware of.

**Comment by paulyoung** [comment]
You can use `lib.fakeSha256` and `lib.fakeSha512`

**Comment by kevincox** [comment]
On Sat., Sep. 5, 2020, 05:36 Paul Young, <notifications@github.com> wrote:

> You can use lib.fakeSha256 and lib.fakeSha512
>
Maybe I'm just being whiney but these aren't very convenient for me.
Notably you need to check if lib is available, if it isn't then you need to
add then remove it. And you need to remember the exact name and
capitalization. It clearly isn't a ton of work, but it is definitely an
inconvenience. Personally I have a macro in my editor which solves the
issue for me but when trying to write tutorials and similar it is not ideal
to say "paste this into your vimrc".

**Comment by NilsIrl** [comment]
What's the equivalent of nix-prefetch-* for `fetchpatch`?

**Comment by cherryblossom000** [comment]
Alternatively, if you already know the revision hash:

```sh
nix-prefetch fetchFromGitHub --owner owner --repo repo --rev 65bb66d364e0d10d00bd848a3d35e2755654655b
The fetcher will be called as follows:
> fetchFromGitHub {
>   owner = "owner";
>   repo = "repo";
>   rev = "65bb66d364e0d10d00bd848a3d35e2755654655b";
>   sha256 = "sha256:0000000000000000000000000000000000000000000000000000";
> }

sha256-8EUDsWeTeZwJNrtjEsUNLMt9I9mjabPRBZG83u7xtPw=
```
